Demo/Program.cs

```diff
@@ -1,26 +1,108 @@
-﻿using External.Encoder;
-using External.Notification;
+﻿using Encoder;
+using Notification;
 using System;
+using System.Text;
 
 namespace Demo
 {
     class Program
     {
+        private static INotificationProvider _notificationProvider;
         static void Main(string[] args)
         {
-            var videoEncoder = new VideoEncoder();
-            var audioEncoder = new AudioEncoder();
-
-            var emailService = new EmailService();
             var smsService = new SmsService();
+            var emailService = new EmailService();
+            var telegramService = new TelegramService();
+            _notificationProvider = new NotificationProvider();
+
+            _notificationProvider.Add(emailService);
+            _notificationProvider.Add(smsService);
+            _notificationProvider.Add(telegramService);
+
+            Console.WriteLine(SourceUploadMessage());
+
+            var sourceType = Console.ReadKey();
+            Console.WriteLine(Environment.NewLine);
+
+            var source = sourceType.KeyChar.GetSource();
+
+            Console.WriteLine(EncoderChooseMessage());
+
+            var encoderType = Console.ReadKey();
+            Console.WriteLine(Environment.NewLine);
+
+            var encoder = encoderType.KeyChar.GetEncoder(source);
+
+            SubscribeToEvents(encoder);
+
+            encoder.Encode();
+        }
+
+        private static void SubscribeToEvents(BaseEncoder encoder)
+        {
+            encoder.Preparing += EncoderOnPreparing;
+            encoder.Starting += EncoderOnStarting;
+            encoder.Finishing += EncoderOnFinishing;
+            encoder.Encoded += EncoderOnEncoded;
+        }
+        private static void EncoderOnPreparing(object sender, EncoderEventArgs args)
+        {
+            // TODO: need to find another way
+            if(sender is int)
+            {
+                var encoder = (IntegerEncoder)sender;
+                if (encoder.Source % 2 != 0)
+                    args.Stop = true;
+            }
 
-            videoEncoder.Encoded += emailService.Send;
-            videoEncoder.Encoded += smsService.Send;
-            audioEncoder.Encoded += emailService.Send;
-            audioEncoder.Encoded += smsService.Send;
+            Console.WriteLine(args.Message);
+        }
+
+        private static void EncoderOnStarting(object sender, EncoderEventArgs args)
+        {
+            Console.WriteLine(args.Message);
+        }
+
+        private static void EncoderOnFinishing(object sender, EncoderEventArgs args)
+        {
+            Console.WriteLine(args.Message);
+        }
```

Comment:
Why do you need 2 different functions that do exactly the same? 
Also, logic inside this function is very simple and I think no need to move it into separate function.